klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,165 @@
+
+def is_subdirectory(haystack, needle):
+    """
+    checks whether `needle` exists as a subdirectory of `haystack`
+    """
+    def func(arg, dirname, fnames):
+        if dirname == needle or needle in fnames:
+            logger.debug('%s is within %s', needle, haystack)
+            raise StopIteration()
+    try:
+        os.path.walk(haystack, func, None)
+    except StopIteration:
+        return True
+    return False
+
+
+class KlausEventHandler(FileSystemEventHandler):
+    """
+    Basic file-system event handler for a klaus app.
+    When git repositories are created, moved or deleted, watchdog will notify us here.
+    We try and update the klaus wsgi app with the modifications as they happen.
+    """
+    def __init__(self, klausapp, *args, **kwargs):
+        self.app = klausapp
+        return super(KlausEventHandler, self).__init__(*args, **kwargs)
+
+    def _add_dir(self, directory):
+        """
+        Attempts to register `directory` with the klaus app as a git repository.
+        If `directory` is not a git repo, this method does nothing.
+        """
+        try:
+            repo = FancyRepo(directory)
+        except dulwich.errors.NotGitRepository:
+            logger.debug("Dulwich says '%s' is not a git repository", directory)
+            return
+
+        matching = filter(lambda existing_repo: existing_repo.name == repo.name, self.app.repos)
+        if matching:
+             logger.warning("skipping repo with duplicate name:%s", repo)
+             return
+        logger.info("Adding repository, '%s' at '%s'", repo.name, directory)
+        self.app.repos.append(repo)
+
+    def on_created(self, event):
+        # We're only interested in directories; A git repo is always a directory
+        if type(event) != DirCreatedEvent:
+            return
+        logger.debug("Received directory creation event for '%s'", event.src_path)
+        self._add_dir(event.src_path)
+        self.app.update_repos_list()
+
+    def on_deleted(self, event):
+        if type(event) != DirDeletedEvent:
+            return
+        matching = filter(lambda existing_repo : existing_repo.path == event.src_path, self.app.repos)
+
+        map(lambda x: self.app.repos.remove(x), matching)
+        logger.info("%s was deleted; removing from klaus", event.src_path)
+        self.app.update_repos_list()
+
+    def on_moved(self, event):
+        """
+        Moved is not the same as deleted, unless the moved dir is no longer a subset of
+        any of the watched trees. In that case we need to consider it a deletion,
+        and remove it from the list of watched repos; otherwise we need to update its path.
+        This is a potentially expensive operation, because the only reliable way
+        to detect whether it is a subset (including symlinks etc) is to walk the path.
+        In practice, this shouldn't affect performance in the general case,
+        as moving a repository is a rare event.
+        """
```

Comment:
Couldn't the symlink problem be solved by simply taking the `os.path.realpath` of the destination path and testing if that real path is a subpath of one of the basedirs? (by string comparison)